### PR TITLE
os/bluestore: introduce hybird allocator

### PR DESCRIPTION
**Pull request by ifed01**
This is a hybrid of AVL and bitmap allocators intended to cap RAM usage for AVL allocator.
New hybrid allocator uses pure AVL allocation until mem usage is below the threshold and fallbacks to bitmap allocator if needed to track short extents there. AVL is still used as a sort of cache to track long extents up to desired RAM cap.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/AvlAllocator.h [hunk 1]
```diff
@@ -43,10 +43,30 @@ struct range_seg_t {
...
+    void operator()(range_seg_t* p)
+    {
+      delete p;
+    }
+  };
+
+protected:
+  /*
+  * ctor intended for the usage from descendant class(es) which
+  * provides handlig for spilled over entries
```
Comment (tchaikov) [inline src/os/bluestore/AvlAllocator.h]: s/handlig/handling/

Comment (ifed01) [reply src/os/bluestore/AvlAllocator.h]: @tchaikov - fixed, thanks!

File: src/os/bluestore/AvlAllocator.cc [hunk 2]
```diff
@@ -55,18 +55,9 @@ uint64_t AvlAllocator::_block_picker(const Tree& t,
...
-  struct dispose_rs {
-    void operator()(range_seg_t* p)
-    {
-      delete p;
-    }
-  };
-}
-
 void AvlAllocator::_add_to_tree(uint64_t start, uint64_t size)
 {
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/AvlAllocator.cc]: done

File: src/common/options.cc [hunk 3]
```diff
@@ -4000,6 +4000,7 @@ std::vector<Option> get_global_options() {
 
     Option("bluefs_allocator", Option::TYPE_STR, Option::LEVEL_DEV)
     .set_default("bitmap")
+    .set_enum_allowed({"bitmap", "stupid", "avl", "hybrid"})
     .set_description(""),
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/common/options.cc]: I like "hybrid" more but open to rename if you find more votes for your variant ;)

**Comment by ifed01** [comment]
@aclamk - extent region merge has been updated as per your suggestion. W.r.t embedding tree directly do you have any rationales for that? IMO this will bring some copy-paste....

File: src/os/bluestore/Allocator.cc [hunk 4]
```diff
@@ -110,6 +115,10 @@ Allocator *Allocator::create(CephContext* cct, string type,
     alloc = new BitmapAllocator(cct, size, block_size, name);
   } else if (type == "avl") {
     return new AvlAllocator(cct, size, block_size, name);
+  } else if (type == "hybrid") {
+    return new HybridAllocator(cct, size, block_size,
+      cct->_conf.get_val<uint64_t>("bluestore_hybrid_alloc_mem_cap"),
+      name);
```
Comment (tchaikov) [inline src/os/bluestore/Allocator.cc:121]: this breaks the build of crimson-osd. could you either add `HybridAllocator.cc` to `src/crimson/os/alienstore/CMakeLists.txt` or disable this part if `WITH_ALIEN` is defined?

Comment (ifed01) [reply src/os/bluestore/Allocator.cc:127]: @tchaikov - added to build.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-04-13_01:52:52-rados-wip-kefu-testing-2020-04-13-0113-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-13_13:31:14-rados-wip-kefu-testing-2020-04-13-1912-distro-basic-smithi/

**Pull request merged**